amacc.c

```diff
@@ -857,6 +854,132 @@ void stmt()
     }
 }
 
+// Ast parsing for IR generating.
+// With a modular code generator, new targets can be easily supported such as native Arm machine code.
+// The compiler now resembles something you could actually build upon (or at least shows the way.)
+void gen(int *n)
+{
+    int i, j, k, l;
+    int *a, *b, *c, *d;
+    i = *n;
+
+    switch (i) {
+    case Num:
+        *++e = IMM; *++e = n[1]; 
+        break;
+    case Loc:
+        *++e = LEA; *++e = n[1];
+        break;
+    case Load:
+        gen(n + 2); 
+        if (n[1] <= INT || n[1] >= PTR) { *++e = (n[1] == CHAR) ? LC : LI; }
+        break;
+    case Assign:
+        gen((int *) n[2]); *++e = PSH; gen(n + 3); *++e = (n[1] == CHAR) ? SC : SI;
+        break;
+    case Inc:
+    case Dec:
+        gen(n + 2);
+        *++e = PSH; *++e = (n[1] == CHAR) ? LC : LI; *++e = PSH;
+        *++e = IMM; *++e = (n[1] >= PTR2) ? sizeof(int) : n[1] >= PTR ? tsize[n[1] - PTR] : 1;
+        *++e = (i == Inc) ? ADD : SUB;
+        *++e = (n[1] == CHAR) ? SC : SI;
+        break;
+    case Cond:
+        gen((int *) n[1]);
+        *++e = BZ; b = ++e;
+        gen((int *) n[2]);
+        if (n[3]) { *b = (int) (e + 3); *++e = JMP; b = ++e; gen((int *) n[3]); }
+        *b = (int)(e + 1);
+        break;
+    case Lor:  gen((int *) n[1]); *++e = BNZ; b = ++e; gen(n + 2); *b = (int)(e + 1); break;
+    case Lan:  gen((int *) n[1]); *++e = BZ;  b = ++e; gen(n + 2); *b = (int)(e + 1); break;
+    case Or:   gen((int *) n[1]); *++e = PSH; gen(n + 2); *++e = OR; break;
+    case Xor:  gen((int *) n[1]); *++e = PSH; gen(n + 2); *++e = XOR; break;
+    case And:  gen((int *) n[1]); *++e = PSH; gen(n + 2); *++e = AND; break;
+    case Eq:   gen((int *) n[1]); *++e = PSH; gen(n + 2); *++e = EQ; break;
+    case Ne:   gen((int *) n[1]); *++e = PSH; gen(n + 2); *++e = NE; break;
+    case Lt:   gen((int *) n[1]); *++e = PSH; gen(n + 2); *++e = LT; break;
+    case Gt:   gen((int *) n[1]); *++e = PSH; gen(n + 2); *++e = GT; break;
+    case Le:   gen((int *) n[1]); *++e = PSH; gen(n + 2); *++e = LE; break;
+    case Ge:   gen((int *) n[1]); *++e = PSH; gen(n + 2); *++e = GE; break;
+    case Shl:  gen((int *) n[1]); *++e = PSH; gen(n + 2); *++e = SHL; break;
+    case Shr:  gen((int *) n[1]); *++e = PSH; gen(n + 2); *++e = SHR; break;
+    case Add:  gen((int *) n[1]); *++e = PSH; gen(n + 2); *++e = ADD; break;
+    case Sub:  gen((int *) n[1]); *++e = PSH; gen(n + 2); *++e = SUB; break;
+    case Mul:  gen((int *) n[1]); *++e = PSH; gen(n + 2); *++e = MUL; break;
+    case Syscall:
+    case Func:
+        c = b = (int *) n[1]; k = 0; l = 1;
+        // how many parameters
+        while (b && l) { ++k; if (!(int *) * b) l = 0; else b = (int *) * b; }
+        j = 0; a = malloc(sizeof(int *) * k); b = c; l = 1;
```

Comment:
[style] Should be `(int *) *k`

Reply:
This is the operator, should we change it?

Reply:
No, it was my mistake.